include/merlin/core_kernels.cuh

```diff
@@ -1966,5 +1968,669 @@ __global__ void dump_kernel(const Table<K, V, M>* __restrict table,
   }
 }
 
+/* If founds[i] = true, read data from corresponding address of table_value_addrs and write to param_values;
+ * if founds[i] = false, write data from param_values to corresponding address of table_value_addrs.
+ * usually called by find_or_insert kernel.
+ */
+template <class V>
+void read_or_write_by_cpu(V** __restrict table_value_addrs, V* __restrict param_values,
+                  const int* __restrict offset, const bool* founds, size_t dim, int N,
+                  int n_worker = 16) {
+  std::vector<std::thread> thds;
+  if (n_worker < 1) n_worker = 1;
+
+  auto functor = [=](V** __restrict table_value_addrs, V* __restrict param_values,
+                    const int* __restrict offset, int handled_size,
+                    int trunk_size) -> void {
+    for (int i = handled_size; i < handled_size + trunk_size; i++) {
+      if (founds[offset[i]] && table_value_addrs[i] != nullptr) {
+        memcpy(param_values + offset[i] * dim, table_value_addrs[i], sizeof(V) * dim);
+      }
+      else {
+        memcpy(table_value_addrs[i], param_values + offset[i] * dim, sizeof(V) * dim);
+      }
+    }
+  };
+
+  size_t trunk_size = N / n_worker;
+  size_t handled_size = 0;
+  for (int i = 0; i < n_worker - 1; i++) {
+    thds.push_back(
+        std::thread(functor, table_value_addrs, param_values, offset, handled_size, trunk_size));
+    handled_size += trunk_size;
+  }
+
+  size_t remaining = N - handled_size;
+  thds.push_back(
+      std::thread(functor, table_value_addrs, param_values, offset, handled_size, remaining));
+
+  for (int i = 0; i < n_worker; i++) {
+    thds[i].join();
+  }
+}
+
+/* find or insert with the end-user specified meta.
+ */
+template <class K, class V, class M, uint32_t TILE_SIZE = 4>
+__global__ void find_or_insert_kernel(const Table<K, V, M>* __restrict table,
+                              const K* __restrict keys, V** __restrict vectors,
```

Comment:
Need to format by following the guide: https://github.com/NVIDIA-Merlin/HierarchicalKV/blob/master/STYLE_GUIDE.md

Reply:
OK, I will do it

Reply:
Nice tool! I learned a lot. Thank you.(^▽^ )